src/Data/Vec/Functional/Properties.agda

```diff
@@ -0,0 +1,265 @@
+updateAt-cong zero    eq xs zero    = eq (xs zero)
+updateAt-cong zero    eq xs (suc j) = refl
+updateAt-cong (suc i) eq xs zero    = refl
+updateAt-cong (suc i) eq xs (suc j) = updateAt-cong i eq (tail xs) j
+
+-- The order of updates at different indices i ≢ j does not matter.
+
+updateAt-commutes : ∀ {n} (i j : Fin n) {f g : A → A} → i ≢ j → (xs : Vector A n) →
+                    updateAt i f (updateAt j g xs) ≗ updateAt j g (updateAt i f xs)
+updateAt-commutes zero    zero    0≢0 xs k       = ⊥-elim (0≢0 refl)
+updateAt-commutes zero    (suc j) _   xs zero    = refl
+updateAt-commutes zero    (suc j) _   xs (suc k) = refl
+updateAt-commutes (suc i) zero    _   xs zero    = refl
+updateAt-commutes (suc i) zero    _   xs (suc k) = refl
+updateAt-commutes (suc i) (suc j) _   xs zero    = refl
+updateAt-commutes (suc i) (suc j) i≢j xs (suc k) = updateAt-commutes i j (i≢j ∘ cong suc) (tail xs) k
+
+------------------------------------------------------------------------
+-- map
+
+map-id : ∀ {n} → (xs : Vector A n) → map id xs ≗ xs
+map-id xs = λ _ → refl
+
+map-cong : ∀ {n} {f g : A → B} → f ≗ g → (xs : Vector A n) → map f xs ≗ map g xs
+map-cong f≗g xs = f≗g ∘ xs
+
+map-∘ : ∀ {n} {f : B → C} {g : A → B} →
+        (xs : Vector A n) → map (f ∘ g) xs ≗ map f (map g xs)
+map-∘ xs = λ _ → refl
+
+lookup-map : ∀ {n} (i : Fin n) (f : A → B) (xs : Vector A n) →
+             map f xs i ≡ f (xs i)
+lookup-map i f xs = refl
+
+map-updateAt : ∀ {n} {f : A → B} {g : A → A} {h : B → B}
+               (xs : Vector A n) (i : Fin n) →
+               f (g (xs i)) ≡ h (f (xs i)) →
+               map f (updateAt i g xs) ≗ updateAt i h (map f xs)
+map-updateAt {n = suc n}       {f = f} xs zero    eq zero    = eq
+map-updateAt {n = suc n}       {f = f} xs zero    eq (suc j) = refl
+map-updateAt {n = suc (suc n)} {f = f} xs (suc i) eq zero    = refl
+map-updateAt {n = suc (suc n)} {f = f} xs (suc i) eq (suc j) = map-updateAt {f = f} (tail xs) i eq j
+
+------------------------------------------------------------------------
+-- _++_
+
+lookup-++-< : ∀ {m n} (xs : Vector A m) (ys : Vector A n) →
+              ∀ i (i<m : toℕ i ℕ.< m) →
+              (xs ++ ys) i ≡ xs (fromℕ< i<m)
+lookup-++-< {m = m} xs ys i i<m = cong Sum.[ xs , ys ] (Finₚ.splitAt-< m i i<m)
+
+lookup-++-≥ : ∀ {m n} (xs : Vector A m) (ys : Vector A n) →
+              ∀ i (i≥m : toℕ i ℕ.≥ m) →
+              (xs ++ ys) i ≡ ys (reduce≥ i i≥m)
+lookup-++-≥ {m = m} xs ys i i≥m = cong Sum.[ xs , ys ] (Finₚ.splitAt-≥ m i i≥m)
+
+lookup-++ˡ : ∀ {m n} (xs : Vector A m) (ys : Vector A n) i →
+             (xs ++ ys) (inject+ n i) ≡ xs i
+lookup-++ˡ {m = m} {n = n} xs ys i = cong Sum.[ xs , ys ] (Finₚ.splitAt-inject+ m n i)
+
+lookup-++ʳ : ∀ {m n} (xs : Vector A m) (ys : Vector A n) i →
+             (xs ++ ys) (raise m i) ≡ ys i
+lookup-++ʳ {m = m} {n = n} xs ys i = cong Sum.[ xs , ys ] (Finₚ.splitAt-raise m n i)
+
+module _ {m} {ys ys′ : Vector A m} where
+
+  ++-cong : ∀ {n} (xs xs′ : Vector A n) →
```

Comment:
This can be generalized to two lengths, I believe. I don't think the proof even needs to change.

Reply:
Can you elaborate please?

Reply:
Aren't there already two lengths? In the module parameters?

Reply:
My mistake, sorry, I was reading the code incorrectly.